File: LeetCode/Explore/Search/MaxAreaOfIsland.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    int maxAreaOfIsland(vector<vector<int>> &grid) {
        int res = 0;
        for (int i = 0; i < grid.size(); i++)
            for (int j = 0; j < grid[0].size(); j++)
                if (grid[i][j] == 1)
                    res = max(res, dfs(grid, i, j));
        return res;
    }

  private:
    int dfs(vector<vector<int>> &grid, int x, int y) {
        if (x < 0 || y < 0 || x >= grid.size() || y >= grid[0].size() ||
            grid[x][y] == 0)
            return 0;
        grid[x][y] = 0;
        int area = 1;
        area += dfs(grid, x + 1, y);
        area += dfs(grid, x - 1, y);
        area += dfs(grid, x, y + 1);
        area += dfs(grid, x, y - 1);
        return area;
    }
};

File: LeetCode/Explore/ByteDance/Array/maxAreaOfIsland.go
package array

func maxAreaOfIsland(grid [][]int) int {
	if len(grid) == 0 {
		return 0
	}
	m, n := len(grid), len(grid[0])

	var dfs func(int, int) int
	dfs = func(i, j int) int {
		if i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == 0 {
			return 0
		}
		res := 1
		grid[i][j] = 0
		res += dfs(i+1, j)
		res += dfs(i-1, j)
		res += dfs(i, j+1)
		res += dfs(i, j-1)
		return res
	}

	res := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 {
				area := dfs(i, j)
				if area > res {
					res = area
				}
			}
		}
	}
	return res
}
